### PR TITLE
Add dotComPublic config for Other gitrepo type

### LINKED ISSUE #56057: dotcom: fix errors for public repos marked as private for other-code-hosts
See https://sourcegraph.com/site-admin/external-services

Image

### PR DESCRIPTION
**Pull request by pjlast**
Closes #56057 

Adds a config option for "Other" code host connections that only works on DotCom:

By adding

```json
"dotComPublic": true
```

to the code host config, repos added by that connection will be marked as public. This needs to be added explicitly, so it is on the configurator of that code host connection that all the repositories added by that connection is in fact public repositories.

https://github.com/sourcegraph/sourcegraph/assets/6427795/01cef46e-2814-43b2-b7be-3572622af2bd

When not on Sourcegraph.com, the config can't be saved at all:

image

## Test plan

Unit test as well as local test.

### CODE REVIEW

File: internal/repos/other.go [hunk 1]
```diff
@@ -71,6 +72,12 @@ func NewOtherSource(ctx context.Context, svc *types.ExternalService, cf *httpcli
 		return nil, err
 	}
 
+	if envvar.SourcegraphDotComMode() {
+		if c.DotComPublic {
+			svc.Unrestricted = true
+		}
+	}
```
Comment (BolajiOlajide) [inline internal/repos/other.go:79]: Could this be further simplified?

```suggestion
	if envvar.SourcegraphDotComMode() && c.DotComPublic {
		svc.Unrestricted = true
	}
```

**Review by BolajiOlajide — APPROVED** [review]

File: internal/repos/other.go (hunk 1 above)
Comment (pjlast) [reply internal/repos/other.go:79]: lol yes, idk why I nested it like that. Thanks!

File: schema/other_external_service.schema.json [hunk 2]
```diff
@@ -80,6 +80,11 @@
           }
         ]
       ]
+    },
+    "dotComPublic": {
```
Comment (mrnugget) [inline schema/other_external_service.schema.json:84]: I'm not 100% happy with the name. Can we make it super descriptive? `makeReposPublicOnDotCom`? I just want this to raise eyebrows when someone comes across it or types it in.

File: internal/repos/other.go [hunk 3]
```diff
@@ -71,6 +72,10 @@ func NewOtherSource(ctx context.Context, svc *types.ExternalService, cf *httpcli
 		return nil, err
 	}
 
+	if envvar.SourcegraphDotComMode() && c.DotComPublic {
+		svc.Unrestricted = true
+	}
```
Comment (mrnugget) [inline internal/repos/other.go:77]: Can we add a validation here to ensure the field can only be set to `true` on dotcom?

https://github.com/sourcegraph/sourcegraph/blob/2b946e6e48ea5bf58f78ab1b34f98cebd5852e8a/internal/database/external_services.go#L442

Comment (pjlast) [reply internal/repos/other.go]: I updated the description with the screenshot, but here it is as well:
image

**Review by mrnugget — APPROVED** [review]

**Review by eseliger — COMMENTED** [review]
nice!

**Pull request merged**